text-editor/text-editor-002-en.md

```diff
@@ -0,0 +1,227 @@
+some broken HTML semantic, so we decided to develop a parser for the received
+HTML markup of the content. The parser will then create React components
+for each tag that we know. Doing so, we only render tags that we're sure
+to support, and we can also easily stylize them.
+
+We wanted to keep the same parsing logic for our text editor as we wanted it
+to be WYSIWYG, ie display the edited content as it would be displayed in the
+public website, and also be able to edit the content.
+
+## The specs
+
+The text editor sould provide the basic fatures of text edition, ie be able to :
+- apply some style to a portion of text (bold, italic, underlined)
+- create an HTTP link
+- create some paragraphs
+- create titles
+
+There are also some media features :
+- be able to insert an image (upload it or pick it from the image collection
+of the company)
+- be able to insert a video (from the video collection of the company)
+- be able to insert a tweet (by specifying the tweet URL)
+- be able to insert a Youtube video (by specifying the video URL)
+
+Such features may seems very common, but they are however tighten to our
+current application as images and videos are dealing with our image and videos
+collection.
+
+This is why such requirements pushed us to develop our own text editor from
+scratch instead of using an existing library.
+
+## Text edition POC
+
+### First attempt : AST
+
+We first wanted to modelize the text to edit using an AST (abstract syntax
+tree). The HTML string was parsed and represented as a tree, in the same way
+the DOM nodes are represented.
+We started to write some simple mutation functions to stylize the text :
+make it bold, underlined, etc... We focused only on the AST part fisrt, the
+rfendering would come later. Unit tests would ensure that the mutations were
+working as expected.
+We managed to update the AST correctly when applying mutations,
+eg the following text node
+
+```js
+{
+    type: "text",
+    text: "lorem ipsum",
+    children: [],
+}
+```
+
+was succesfully embed into an other node when we wanted to make it bold :
+
+```js
+{
+    type: "b",
+    text: "",
+    children: [
+        {
+            type: "text",
+            text: "lorem ipsum",
+            children: []
+        }
+    ]
+}
+```
+
+However, things were getting way more complicated when we wanted to update
+some selected text that were starting in a paragraph and ending on an other,
+or also when the selection was overlapsing on different tags, such as
+
+```
+<b>Hello <em>John Doe</em></b>, how are you ?
+    |---------------|
+selection start   selection end
+```
+
+We realized that applying such changes on the AST would have beed difficult
+for us. In the mean time, the developer who started to work on the AST was
+required to work on an other project so we decided not to use this AST anymore
+as it would have been very time consuming to develop and maintain (and also
+the learning surve was not easy for newcomers)
```

Comment:
Maybe emphasis here on the FP paradigm and why it was difficult for us to follow along ?